galley/pkg/probe/controller.go

```diff
@@ -0,0 +1,88 @@
+package probe
+
+import (
+	"time"
+	"io/ioutil"
+
+	"istio.io/istio/pkg/probe"
+	"istio.io/istio/pkg/log"
+
+)
+const(
+     DefaultProbeCheckInterval = 2 * time.Second
+     DefaultLivenessProbeFilePath = "/healthLiveness"
+     DefaultReadinessProbeFilePath = "/healthReadiness"
+)
+type ProbeCheckController struct {
+	livenessProbe *probe.Probe
+	readinessProbe *probe.Probe
+	livenessProbeOptions *probe.Options
+	readinessProbeOptions *probe.Options
+
+}
+
+
+func NewProbeCheckController(livenessProbeOptions *probe.Options, readinessProbeOptions *probe.Options) (*ProbeCheckController) {
+	var livenessProbe, readinessProbe *probe.Probe
+	if livenessProbeOptions != nil {
+		livenessProbe = probe.NewProbe()
+		livenessProbeController := probe.NewFileController(livenessProbeOptions)
+		livenessProbe.RegisterProbe(livenessProbeController, "liveness")
+		livenessProbeController.Start()
+		livenessProbe.SetAvailable(nil)
+	}
+	if readinessProbeOptions != nil {
+		readinessProbe = probe.NewProbe()
+		readinessProbeController := probe.NewFileController(readinessProbeOptions)
+		readinessProbe.RegisterProbe(readinessProbeController, "readiness")
+		readinessProbeController.Start()
+		readinessProbe.SetAvailable(nil)
+	}
+
```

Comment:
Can we align this with how [mixer uses the readiness probe package](https://github.com/istio/istio/blob/ced35e7047ea69b92a97a30eba44fba7d7fee516/mixer/pkg/server/server.go#L224-L234)? Specifically, the probe package itself seems to already handle updating the health file. We only need to invoke `SetAvailable(<status>)` based on server and validation readiness.

Reply:
Mixer have not used this part actually because they have not defined liveness and readiness probe options in their helm/docker file.  the security part is relying on this mechanism now: https://github.com/istio/istio/blob/master/security/pkg/probe/controller.go , so we are now under the same way with it.

Reply:
currently I have used the same logic to detect the liveness and readiness for galley:
func (c *ProbeCheckController) checkLiveness() error {
	content := []byte(`ok`)
	if err := ioutil.WriteFile(c.livenessProbeOptions.Path, content, 0644); err != nil {
		log.Errorf("Health check update of %q failed: %v", c.livenessProbeOptions.Path, err)
		return err

	}
	return nil
}


func (c *ProbeCheckController) checkReadiness() error {
	content := []byte(`ok`)
	if err := ioutil.WriteFile(c.readinessProbeOptions.Path, content, 0644); err != nil {
		log.Errorf("Health check update of %q failed: %v", c.readinessProbeOptions.Path, err)
		return err

	}
	return nil
}

So in case there is error, the probe package itself will get the error and remove the heath file. 

So is above logic valid to check the galley liveness and readiness? (Just copy the logic from validation code)?

Reply:
See https://github.com/istio/istio/pull/7986 (release-1.0 branch). This change should also be merged into master. This decouples the config reconciliation loop from the health check updates. To do this, I used the `istio.io/pkg/probe` package instead of writing the file directly. We should use the same mechanism for the galley service health checks.

Reply:
Done~~ Thanks for the clue 👍